pkg/ipmasq/ipmasq.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipmasq
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/maps/ipmasq"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "ipmasq")
+)
+
+// ipnet is a wrapper type for net.IPNet to enable de-serialization of CIDRs
+type Ipnet net.IPNet
+
+func (c *Ipnet) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var str string
+	if err := unmarshal(&str); err != nil {
+		return err
+	}
+
+	_, n, err := net.ParseCIDR(strings.Trim(str, `"`))
+	if err != nil {
+		return fmt.Errorf("Invalid CIDR %s: %s", str, err)
+	}
+	*c = Ipnet(*n)
+	return nil
+
+}
+
+// config represents the ip-masq-agent configuration file encoded as YAML
+type config struct {
+	NonMasqCIDRs []Ipnet `yaml:"nonMasqueradeCIDRs"`
+}
+
+// IPMasqMap is an interface describing methods for manipulating an ipmasq map
+type IPMasqMap interface {
+	Update(cidr net.IPNet) error
+	Delete(cidr net.IPNet) error
+	Dump() ([]net.IPNet, error)
+}
+
+// IPMasqAgent represents a state of the ip-masq-agent
+type IPMasqAgent struct {
+	configPath             string
+	nonMasqCIDRsFromConfig map[string]net.IPNet
+	nonMasqCIDRsInMap      map[string]net.IPNet
+	ipMasqMap              IPMasqMap
+}
+
+// Start starts the "ip-masq-agent" controller which is used to sync the ipmasq
+// BPF maps.
+func Start(configPath string, syncPeriod time.Duration) error {
+	return start(configPath, syncPeriod, &ipmasq.IPMasqBPFMap{}, controller.NewManager())
+}
+
+func start(configPath string, syncPeriod time.Duration,
+	ipMasqMap IPMasqMap, manager *controller.Manager) error {
+
+	a := &IPMasqAgent{
+		configPath:        configPath,
+		nonMasqCIDRsInMap: map[string]net.IPNet{},
+		ipMasqMap:         ipMasqMap,
+	}
+
+	if err := a.restore(); err != nil {
+		return err
+	}
+
+	manager.UpdateController("ip-masq-agent",
+		controller.ControllerParams{
+			DoFunc: func(ctx context.Context) error {
+				return a.Update()
```

Comment:
Did you consider signing up for filesystem events for update of the file rather than polling on a regular basis?

I guess it shouldn't matter that much though, I presume we expect these to be relatively small and not take a lot of time to process.

Reply:
Yep, an ip-masq-agent config file is usually very simple and small, so therefore I was reluctant to use fsnotify.